pkg/podrestartmonitor/pod_restart_monitor.go

```diff
@@ -0,0 +1,183 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package podrestartmonitor
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"time"
+
+	"github.com/golang/glog"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+
+	"k8s.io/node-problem-detector/pkg/problemdaemon"
+	"k8s.io/node-problem-detector/pkg/types"
+)
+
+const PodRestartMonitorName = "pod-restart-monitor"
+
+func init() {
+	problemdaemon.Register(
+		PodRestartMonitorName,
+		types.ProblemDaemonHandler{
+			CreateProblemDaemonOrDie: NewMonitorOrDie,
+			CmdOptionDescription:     "Set to config file paths."})
+}
+
+type podRestartMonitor struct {
+	configPath string
+	config     MonitorConfig
+	stop       chan struct{}
+	clientset  *kubernetes.Clientset
+	statuses   chan *types.Status
+}
+
+// NewMonitorOrDie create a new podRestartMonitor, panic if error occurs.
+func NewMonitorOrDie(configPath string) types.Monitor {
+	l := &podRestartMonitor{
+		configPath: configPath,
+	}
+
+	f, err := ioutil.ReadFile(configPath)
+	if err != nil {
+		glog.Fatalf("Failed to read configuration file %q: %v", configPath, err)
+	}
+	err = json.Unmarshal(f, &l.config)
+	if err != nil {
+		glog.Fatalf("Failed to unmarshal configuration file %q: %v", configPath, err)
+	}
+
+	// Apply default configurations
+	if l.config.CheckInterval == 0 {
+		l.config.CheckInterval = Duration(5 * time.Minute)
+	}
+	if l.config.RestartThreshold == 0 {
+		l.config.RestartThreshold = 5
+	}
+	if l.config.Namespace == "" {
+		l.config.Namespace = "kube-system"
+	}
+	if l.config.PodSelector == "" {
+		l.config.PodSelector = "k8s-app=aws-node"
+	}
+
+	glog.Infof("Finish parsing pod restart monitor config file %s: %+v", l.configPath, l.config)
+	l.stop = make(chan struct{})
+
+	return l
+}
+
+func (l *podRestartMonitor) Start() (<-chan *types.Status, error) {
+	glog.Infof("Start pod restart monitor %s", l.configPath)
+
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	l.clientset, err = kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	l.statuses = make(chan *types.Status)
+
+	go l.monitorLoop()
+
+	return l.statuses, nil
+}
+
+func getNodeName() string {
+	return os.Getenv("NODE_NAME")
+}
+
+func (l *podRestartMonitor) Stop() {
+	glog.Infof("Stopping monitor %s", l.configPath)
+	close(l.stop)
+}
+
+const ConditionTooManyRestarts = "TooManyRestarts"
+
+// monitorLoop is the main loop of log monitor.
+func (l *podRestartMonitor) monitorLoop() {
+	nodeName := getNodeName()
+MainLoop:
+	for {
+		select {
+		case <-l.stop:
+			glog.Warningf("monitor stopped")
+			return
+		case <-time.After(time.Duration(l.config.CheckInterval)):
+			list, err := l.clientset.CoreV1().Pods(l.config.Namespace).List(metav1.ListOptions{
+				LabelSelector: l.config.PodSelector,
```

Comment:
I believe it will be an empty list. Try doing `kubectl get pod -l fakelabel=fakevalue -o json` in your cluster, you will get this:

```json
{
    "apiVersion": "v1",
    "items": [],
    "kind": "List",
    "metadata": {
        "resourceVersion": "",
        "selfLink": ""
    }
}
```

So the loop will just do nothing, I think. It might trip up some people who get their namespace settings wrong or typo the selector when configuring their checks, but it would be consistent with how `kubectl` works.

Reply:
That sounds good. I was worried that this might return an error, but an empty list is a handleable output.

For misconfigured selectors, we could have a warning log. But I don't see much value, I think it's good as is.